app/src/main/java/mozilla/lockbox/presenter/AutofillFilterPresenter.kt

```diff
@@ -7,47 +7,84 @@
-//    val itemSelection: Observable<ItemViewModel>
-//    fun updateItems(items: List<ItemViewModel>)
+    val filterTextEntered: Observable<CharSequence>
+    val filterText: Consumer<in CharSequence>
+    val cancelButtonClicks: Observable<Unit>
+    val cancelButtonVisibility: Consumer<in Boolean>
+    val itemSelection: Observable<ItemViewModel>
+    fun updateItems(items: List<ItemViewModel>, displayNoEntries: Boolean)
 }
 
+@ExperimentalCoroutinesApi
 class AutofillFilterPresenter(
     val view: AutofillFilterView,
     val dispatcher: Dispatcher = Dispatcher.shared,
     val dataStore: DataStore = DataStore.shared
 ) : Presenter() {
     override fun onViewReady() {
-        view.fillMeButtonClicks
+        view.onDismiss
+            .map { AutofillAction.Cancel }
+            .subscribe(dispatcher::dispatch)
+            .addTo(compositeDisposable)
+
+        val itemViewModelList = dataStore.list.mapToItemViewModelList()
+
+        Observables.combineLatest(view.filterTextEntered, itemViewModelList)
+            .filterItemsForText()
             .map {
-                AutofillAction.Complete(
-                    ServerPassword("bllah",
-                        "www.blah.com",
-                        "cats@cats.com",
-                        "dawgzone")
-                )
+                val displayNoEntries = it.second.isEmpty() || !it.first.isEmpty()
+                val itemList = if (it.first.isEmpty()) emptyList() else it.second
+                Pair(itemList, displayNoEntries)
             }
-            .subscribe(dispatcher::dispatch)
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe { view.updateItems(it.first, it.second) }
             .addTo(compositeDisposable)
 
-        view.onDismiss
+        view.filterTextEntered
+            .map { it.isNotEmpty() }
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(view.cancelButtonVisibility)
+            .addTo(compositeDisposable)
+
+        view.cancelButtonClicks
+            .map { "" }
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(view.filterText)
+            .addTo(compositeDisposable)
+
+        view.itemSelection
+            .switchMap { dataStore.get(it.guid) }
             .map {
+                it.value?.let { serverPassword ->
+                    return@map AutofillAction.Complete(serverPassword)
+                }
+
                 AutofillAction.Cancel
             }
             .subscribe(dispatcher::dispatch)
             .addTo(compositeDisposable)
     }
+
+    private fun Observable<Pair<CharSequence, List<ItemViewModel>>>.filterItemsForText(): Observable<Pair<CharSequence, List<ItemViewModel>>> {
+        return this.map { pair ->
+            Pair(pair.first, pair.second.filter {
+                it.title.contains(pair.first, true) ||
+                    it.subtitle.contains(pair.first, true)
+            })
```

Comment:
Nit: `Pair(pair.pair, pair.filter { it.contains pair }`

This feels like it could be clearer.

```kotlin
return this.map { (text, list) -> 
    val filtered = list.filter { it.title.contains(text, true) || it.subtitle.contains(text, true) }
    text to filtered
}
```

Reply:
Also: I'm not sure why this needs its own extension method.

Reply:
Also: consider adding a `contains` method to `ItemViewModel`.